### PR TITLE
FLIP-2243: Update runJobs lambda

### PR DESCRIPTION
**Pull request by MichaelOSullivanAnswer**
This PR includes changes to enable the selection and configuration of a predefined NVFlare aggregator. The following snapshots show the `config.json` supplied for a successful training run and the logs outputted to the UI from the `runJobs` lambda: 

image

image

image

The following images show the `config.json` and associated error with some invalid config options:

A string passed in as a weight
image
image

A number outside the accepted range
image
image

Invalid type for AGGREGATION_WEIGHTS
image
image

Invalid aggregator
image
image

Invalid trust selected for weights
image
image

### CODE REVIEW

File: services/fl-service/src/interfaces.ts [hunk 1]
```diff
@@ -14,6 +14,8 @@
  *  limitations under the License.
  */
 
+import { Dictionary } from "underscore";
+
```
Comment (mbaltrimas) [inline services/fl-service/src/interfaces.ts:18]: i think this is unused

File: services/fl-service/src/services/fl-service.ts [hunk 2]
```diff
@@ -75,10 +78,57 @@ const validateConfig = (config: IOverridableConfig): IOverridableConfig => {
...
+        if (FLAggregators.includes(config.AGGREGATOR)) {
+            validatedConfig.AGGREGATOR = config.AGGREGATOR;
+        } else {
+
+            throw new Error(`An unknown aggregator has been selected: ${config.AGGREGATOR}`);
+        }
+    }
+
+    if (config.AGGREGATION_WEIGHTS) {
+        if (config.AGGREGATION_WEIGHTS.constructor !== Object) {
```
Comment (handyandypandyandy) [inline services/fl-service/src/services/fl-service.ts:95]: I've not come across this before but fair enough if it works! just to keep consistency though, could you use
```suggestion
        if (typeof config.AGGREGATION_WEIGHTS !== "object") {
```

File: services/fl-service/src/services/fl-service.ts [hunk 3]
```diff
@@ -75,10 +78,57 @@ const validateConfig = (config: IOverridableConfig): IOverridableConfig => {
...
-    if(typeof config.IGNORE_RESULT_ERROR === "boolean") {
+    if (typeof config.IGNORE_RESULT_ERROR === "boolean") {
         validatedConfig.IGNORE_RESULT_ERROR = config.IGNORE_RESULT_ERROR;
     }
 
+    if (config.AGGREGATOR) {
+        if (FLAggregators.includes(config.AGGREGATOR)) {
+            validatedConfig.AGGREGATOR = config.AGGREGATOR;
+        } else {
+
```
Comment (handyandypandyandy) [inline services/fl-service/src/services/fl-service.ts:89]: so many empty lines throughout this function
```suggestion
```

File: services/fl-service/src/interfaces.ts [hunk 4]
```diff
@@ -14,6 +14,8 @@
  *  limitations under the License.
  */
 
+import { Dictionary } from "underscore";
```
Comment (handyandypandyandy) [inline services/fl-service/src/interfaces.ts:17]: it doesn't seem like this is being used anywhere

File: services/fl-service/src/services/fl-service.ts [hunk 5]
```diff
@@ -75,10 +78,57 @@ const validateConfig = (config: IOverridableConfig): IOverridableConfig => {
...
+    if (config.AGGREGATION_WEIGHTS) {
+        if (config.AGGREGATION_WEIGHTS.constructor !== Object) {
+
+            throw new Error("Value supplied for AGGREGATION_WEIGHTS not a valid JSON object");
+        }
+        
+        const trusts = Object.keys(config.AGGREGATION_WEIGHTS);
+        const weights = Object.values(config.AGGREGATION_WEIGHTS);
+
+        trusts.map(trust => {
```
Comment (handyandypandyandy) [inline services/fl-service/src/services/fl-service.ts:103]: This validation step can be removed. The type you've specified for the aggregation weights wont serialise any values other than KCH and UCLH

File: services/fl-service/src/services/fl-service.ts [hunk 6]
```diff
@@ -75,10 +78,57 @@ const validateConfig = (config: IOverridableConfig): IOverridableConfig => {
...
+            }
+        });
+
+        weights.map(weight => {
+            logger.info(`Validating aggregation weight value: ${weight}.`);
+
+            if (
+                typeof weight === "number" &&
+                !isNaN(weight) &&
+                weight >= 0 &&
```
Comment (handyandypandyandy) [inline services/fl-service/src/services/fl-service.ts:119]: could you add these as constants `MinimumAggregationWeight` & `MaximumAggregationWeight`?

File: services/fl-service/src/interfaces.ts [hunk 7]
```diff
@@ -56,6 +58,11 @@ export interface IStartTrainingInputPayload {
     globalRounds: number;
     trusts: string[];
     ignoreResultError?: boolean;
+    aggregator?: string;
+    aggregationWeights?: {
+        KCH: number;
+        UCLH: number;
+    };
```
[the remark this answers is not in the document]
Comment (MichaelOSullivanAnswer) [reply services/fl-service/src/interfaces.ts:65]: With this change does that mean I should leave in the trust validation?

File: services/fl-service/src/services/fl-service.ts (hunk 2 above)
Comment (MichaelOSullivanAnswer) [reply services/fl-service/src/services/fl-service.ts:95]: This issue with that implementation is that it won't catch an array if it is passed in.

Comment (handyandypandyandy) [reply services/fl-service/src/services/fl-service.ts:94]: yep fair enough, just tested and you're right

File: services/fl-service/src/interfaces.ts [hunk 8]
```diff
@@ -175,4 +183,21 @@ export interface IOverridableConfig {
...
 }
+
+export const FLAggregators: string[] = [
+    "InTimeAccumulateWeightedAggregator", "AccumulateWeightedAggregator"
+];
+
+export const AvailableTrusts: string[] = [
+    "KCH", "UCLH"
+];
```
[the remark this answers is not in the document]
Comment (MichaelOSullivanAnswer) [reply services/fl-service/src/interfaces.ts]: You're right, has been removed

**Review by mbaltrimas — DISMISSED** [review]

File: services/fl-service/src/services/fl-service.ts [hunk 9]
```diff
@@ -75,10 +77,42 @@ const validateConfig = (config: IOverridableConfig): IOverridableConfig => {
...
+    }
+
+    if (config.AGGREGATION_WEIGHTS) {
+        if (config.AGGREGATION_WEIGHTS.constructor !== Object) {
+            throw new Error("Value supplied for AGGREGATION_WEIGHTS not a valid JSON object");
+        }
+        
+        const weights = Object.values(config.AGGREGATION_WEIGHTS);
+
+        weights.map(weight => {
```
Comment (handyandypandyandy) [inline services/fl-service/src/services/fl-service.ts:99]: As the return value of `map()` is not being assigned to anything, it becomes redundant. Use `weights.forEach(weight => ...` instead

**Review by handyandypandyandy — APPROVED** [review]

**Pull request merged**